File: lib/src/ast/selector/type.dart
// Copyright 2016 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:meta/meta.dart';
import 'package:source_span/source_span.dart';

import '../../extend/functions.dart';
import '../../visitor/interface/selector.dart';
import '../selector.dart';

/// A type selector.
///
/// This selects elements whose name equals the given name.
///
/// {@category AST}
final class TypeSelector extends SimpleSelector {
  /// The element name being selected.
  final QualifiedName name;

  @override
  int get specificity => 1;

  TypeSelector(this.name, FileSpan span) : super(span);

  @override
  T accept<T>(SelectorVisitor<T> visitor) => visitor.visitTypeSelector(this);

  /// @nodoc
  @override
  @internal
  TypeSelector addSuffix(String suffix) => TypeSelector(
        QualifiedName(name.name + suffix, namespace: name.namespace),
        span,
      );

  /// @nodoc
  @override
  @internal
  List<SimpleSelector>? unify(List<SimpleSelector> compound) {
    if (compound.firstOrNull case UniversalSelector() || TypeSelector()) {
      var unified = unifyUniversalAndElement(this, compound.first);
      if (unified == null) return null;
      return [unified, ...compound.skip(1)];
    } else {
      return [this, ...compound];
    }
  }

  @override
  bool isSuperselector(SimpleSelector other) =>
      super.isSuperselector(other) ||
      (other is TypeSelector &&
          name.name == other.name.name &&
          (name.namespace == '*' || name.namespace == other.name.namespace));

  @override
  bool operator ==(Object other) => other is TypeSelector && other.name == name;

  @override
  int get hashCode => name.hashCode;
}

File: pkg/sass-parser/lib/src/selector/type.ts
// Copyright 2025 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import * as postcss from 'postcss';

import {LazySource} from '../lazy-source';
import type {AnyNode, NodeProps} from '../node';
import type {AnyStatement} from '../statement';
import * as sassInternal from '../sass-internal';
import * as utils from '../utils';
import {QualifiedName, QualifiedNameProps} from './qualified-name';
import {SimpleSelector} from './index';

/**
 * The initializer properties for {@link TypeSelector}.
 *
 * @category Selector
 */
export interface TypeSelectorProps extends NodeProps {
  type: QualifiedName | QualifiedNameProps;
  raws?: TypeSelectorRaws;
}

/**
 * Raws indicating how to precisely serialize an {@TypeSelector}.
 *
 * @category Selector
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface -- No raws for a type selector yet.
export interface TypeSelectorRaws {}

/**
 * A type selector.
 *
 * This selects elements of the given type.
 *
 * @category Selector
 */
export class TypeSelector extends SimpleSelector {
  readonly sassType = 'type' as const;
  declare raws: TypeSelectorRaws;

  /** The class name that this selects. */
  get type(): QualifiedName {
    return this._type;
  }
  set type(type: QualifiedName | QualifiedNameProps) {
    if (this._type) this._type.parent = undefined;
    const built =
      typeof type === 'object' &&
      'sassType' in type &&
      type.sassType === 'qualified-name'
        ? type
        : new QualifiedName(type);
    built.parent = this;
    this._type = built;
  }
  private declare _type: QualifiedName;

  constructor(defaults: TypeSelectorProps);
  /** @hidden */
  constructor(_: undefined, inner: sassInternal.TypeSelector);
  constructor(defaults?: object, inner?: sassInternal.TypeSelector) {
    super(defaults);
    if (inner) {
      this.source = new LazySource(inner);
      this.type = new QualifiedName(undefined, inner.name);
    }
  }

  clone(overrides?: Partial<TypeSelectorProps>): this {
    return utils.cloneNode(this, overrides, ['raws', 'type']);
  }

  toJSON(): object;
  /** @hidden */
  toJSON(_: string, inputs: Map<postcss.Input, number>): object;
  toJSON(_?: string, inputs?: Map<postcss.Input, number>): object {
    return utils.toJSON(this, ['type'], inputs);
  }

  /** @hidden */
  toString(): string {
    return this.type.toString();
  }

  /** @hidden */
  get nonStatementChildren(): ReadonlyArray<Exclude<AnyNode, AnyStatement>> {
    return [this.type];
  }
}
